### PR TITLE
listener: original_dst filter config api

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
Make original_dst listener filter configurable.

Signed-off-by: Yuchen Dai <silentdai@gmail.com>
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by lambdai** [comment]
@davinci26 I think DNAT method should work in windows as it should follow the current flow.

I can imagine you can create a new method for windows, *probably* save a syscall?

**Review by davinci26 — COMMENTED** [review]
`SO_ORIGINAL_DST` works on Windows, but we should #ifdef it because it is not available on older releases

File: api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto [hunk 1]
```diff
@@ -17,4 +17,16 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 message OriginalDst {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.listener.original_dst.v2.OriginalDst";
+  enum Method {
+    // Use sockopt SO_ORIGINAL_DST or IP6T_SO_ORIGINAL_DST to recover the original local address. This is the default method.
+    Dnat = 0;
```
Comment (htuch) [inline api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:22]: `DNAT`, etc.

Comment (htuch) [reply api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:22]: I'm actually confused about this feature as a user. How and when do I set these options? It looks like there is more than just recovery, e.g. "mark the local address as restored"..

Comment (lambdai) [reply api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:22]: > It looks like there is more than just recovery, e.g. "mark the local address as restored"
Exactly! DNAT will do both "recover from sockopt" and "mark as restored". That's the current filter's behavior. 
I just name it DNAT...

Inspired by the @tomastigera 's experimental commit in #16130 
In TPROXY-and-no-DNAT iptable rules, the original_dst_cluster need the latter "mark as restored" flag. That's what the second method does.
Some integration tests will also find this behavior useful.

For anybody else, either explicitly set DNAT or use the default value DNAT.

Comment (htuch) [reply api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:22]: OK, but "as a reader" I have no idea how to use this still. This is user facing documentation and needs to spell out what these features are and link to references and/or Envoy documentation as appropriate.

**Comment by lambdai** [comment]
> `SO_ORIGINAL_DST` works on Windows, but we should #ifdef it because it is not available on older releases

Thank you for the clarification! That says we don't need extra api in proto for windows.

File: api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto (hunk 1 above)
Comment (lambdai) [reply api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:22]: Make sense to me. WIll update

File: api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto [hunk 2]
```diff
@@ -17,4 +17,16 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
+
+    // Mark the local address as restored so that the :ref:`original_dst cluster<envoy_v3_api_enum_value_config.cluster.v3.Cluster.DiscoveryType.ORIGINAL_DST>`
+    // can be utilized. This recovery method is used when TPROXY rule is used without DNAT rule.
+    CurrentLocalAddress = 1;
+  }
+
+  // The method to recover the original local address. The default method is Dnat if the field is not set.
+  // [#not-implemented-hide:]
+  Method address_recovery_method = 1;
```
Comment (tomastigera) [inline api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:31]: Should the default be Dnat?

```suggestion
  Method address_recovery_method = Dnat;
```

File: api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto [hunk 3]
```diff
@@ -17,4 +17,16 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 message OriginalDst {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.listener.original_dst.v2.OriginalDst";
+  enum Method {
+    // Use sockopt SO_ORIGINAL_DST or IP6T_SO_ORIGINAL_DST to recover the original local address. This is the default method.
+    Dnat = 0;
+
+    // Mark the local address as restored so that the :ref:`original_dst cluster<envoy_v3_api_enum_value_config.cluster.v3.Cluster.DiscoveryType.ORIGINAL_DST>`
+    // can be utilized. This recovery method is used when TPROXY rule is used without DNAT rule.
```
Comment (tomastigera) [inline api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:25]:
```suggestion
    // can be utilized. This recovery method should be used with TPROXY redirection and transparent sockets unless in the unlikely case when the user whats to recover a possibly pre-DNAT address.
```

File: api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto (hunk 1 above)
Comment (tomastigera) [reply api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:22]: I think this does not need to dive into the SO_ORIGINAL_DST detail and rather stress that this is the method that needs to be used with REDIRECT redirection.

**Review by tomastigera — COMMENTED** [review]
Some wording clarification for consideration. Thanks @lambdai for doing this!

File: api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto [hunk 4]
```diff
@@ -17,4 +17,17 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 message OriginalDst {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.listener.original_dst.v2.OriginalDst";
+
+  enum Method {
+    // Use sockopt SO_ORIGINAL_DST or IP6T_SO_ORIGINAL_DST to recover the original local address from the DNAT or
+    // REDIRECT iptables target. This is the default method.
+    DNAT = 0;
```
Comment (antoniovicente) [inline api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:24]: Protobuf best practices call for the 0th enum value being reserved for DEFAULT or NOT_SET.  Would you consider adding a separate enum value for DEFAULT?

File: api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto (hunk 2 above)
Comment (antoniovicente) [reply api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:32]: The "= 1" refers to the field number.  The default value of enum fields is 0, so effectively the default is DNAT.  Do see my comment above about proto best practices using 0 for an enum value with "default" in its name.

File: source/extensions/filters/listener/original_dst/config.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+class Config {
+public:
+  Config() = default;
+  Config(const envoy::extensions::filters::listener::original_dst::v3::OriginalDst& config,
+         const envoy::config::core::v3::TrafficDirection& traffic_direction)
+      : method_(getOriginalDstMethod(config)), traffic_direction_(traffic_direction) {}
+
+  enum class OriginalDstMethod {
+    SocketOption = 0,
+    NoOp = 1,
```
Comment (antoniovicente) [inline source/extensions/filters/listener/original_dst/config.h:21]: Please change `NoOp` to `CurrentLocalAddress` for consistency.  It would be also good to document the meaning of these values.

The disparity between `DNAT` in the proto definition vs `SocketOption` in this enum is also odd.  It would be good to reconcile naming.

File: source/extensions/filters/listener/original_dst/config.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  };
+  OriginalDstMethod method_{OriginalDstMethod::SocketOption};
+  envoy::config::core::v3::TrafficDirection traffic_direction_;
```
Comment (antoniovicente) [inline source/extensions/filters/listener/original_dst/config.h:24]: Could you make these data members private or change the type of this object from `class` to `struct`?

Is it possible to mark them `const`?

File: source/extensions/filters/listener/original_dst/original_dst.h [hunk 7]
```diff
@@ -14,20 +15,15 @@ namespace OriginalDst {
...
+  explicit OriginalDstFilter(const Config& config) : config_(config) {}
 
   virtual Network::Address::InstanceConstSharedPtr getOriginalDst(Network::Socket& sock);
 
   // Network::ListenerFilter
   Network::FilterStatus onAccept(Network::ListenerFilterCallbacks& cb) override;
 
 private:
-  envoy::config::core::v3::TrafficDirection traffic_direction_;
+  Config config_;
```
Comment (antoniovicente) [inline source/extensions/filters/listener/original_dst/original_dst.h:26]: nit: `const Config config_;`

**Review by antoniovicente — COMMENTED** [review]
@htuch could you look at the API?  I think most of the feedback relevant to this PR is related to that.

Looks good in general, just some naming consistency nits.

File: api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto (hunk 4 above)
Comment (htuch) [reply api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:24]: In this case, we're basically saying "client-specific default, set explicit DNAT if that is what you really want". This is consistent with API guidelines for multi-client.

File: api/BUILD [hunk 8]
```diff
@@ -133,12 +133,18 @@ proto_library(
...
         "//envoy/config/endpoint/v3:pkg",
+        "//envoy/config/filter/thrift/router/v2alpha1:pkg",
         "//envoy/config/grpc_credential/v3:pkg",
+        "//envoy/config/health_checker/redis/v2:pkg",
         "//envoy/config/listener/v3:pkg",
         "//envoy/config/metrics/v3:pkg",
         "//envoy/config/overload/v3:pkg",
         "//envoy/config/ratelimit/v3:pkg",
         "//envoy/config/rbac/v3:pkg",
+        "//envoy/config/resource_monitor/fixed_heap/v2alpha:pkg",
```
Comment (htuch) [inline api/BUILD]: What is this unrelated change?

File: api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto (hunk 4 above)
Comment (htuch) [reply api/envoy/extensions/filters/listener/original_dst/v3/original_dst.proto:24]: Yes, that is what I was saying.

**Comment by antoniovicente** [comment]
See open comments.

/wait

**Comment by wangfakang** [comment]
hello @lambdai  have any plans to update this PR? I also meet this [issue](https://github.com/envoyproxy/envoy/issues/16130) mentioned by @tomastigera.
